parser_library/src/lsp/lsp_context.cpp

```diff
@@ -75,6 +77,436 @@ hover_result hover_text(const variable_symbol_definition& sym)
 }
 } // namespace
 
+const std::unordered_map<context::symbol_origin, document_symbol_kind> document_symbol_item_kind_mapping_symbol {
+    { context::symbol_origin::DAT, document_symbol_kind::DAT },
+    { context::symbol_origin::EQU, document_symbol_kind::EQU },
+    { context::symbol_origin::MACH, document_symbol_kind::MACH },
+    { context::symbol_origin::UNKNOWN, document_symbol_kind::UNKNOWN }
+};
+
+const std::unordered_map<context::section_kind, document_symbol_kind> document_symbol_item_kind_mapping_section {
+    { context::section_kind::COMMON, document_symbol_kind::COMMON },
+    { context::section_kind::DUMMY, document_symbol_kind::DUMMY },
+    { context::section_kind::EXECUTABLE, document_symbol_kind::EXECUTABLE },
+    { context::section_kind::READONLY, document_symbol_kind::READONLY }
+};
+
+const std::unordered_map<occurence_kind, document_symbol_kind> document_symbol_item_kind_mapping_macro {
+    { occurence_kind::VAR, document_symbol_kind::VAR }, { occurence_kind::SEQ, document_symbol_kind::SEQ }
+};
+
+context::id_index lsp_context::find_macro_copy_id(const context::processing_stack_t& stack, int i) const
+{
+    assert(i != 0);
+    return stack[i].member_name == opencode_->hlasm_ctx.ids().empty_id ? &stack[i].proc_location.file
+                                                                       : stack[i].member_name;
+}
+
+document_symbol_list_s lsp_context::document_symbol_macro(const std::string& document_uri) const
+{
+    auto copy_occs = copy_occurences(document_uri);
+    document_symbol_list_s result;
+    for (const auto& [def, info] : macros_)
+    {
+        if (def->definition_location.file == document_uri)
+        {
+            for (const auto& var : info->var_definitions)
+            {
+                if (find_occurence_with_scope(document_uri, var.def_position).first == nullptr)
+                {
+                    modify_with_copy(result, var.name, copy_occs, document_symbol_kind::VAR);
```

Comment:
If you mean the loop from line 114 to 125, then it is at least O(voqrc), where
v is the number of VAR symbols
o is the number of copy files
q is the number of occurrences in a copy file
r is the number of document_symbol_items already in result
c is the number of children in an document_symbol_item

In other words it is really slow. The reason why it is like that is that the variable symbols of copy files are together with the variable symbols in macro file in one vardef_storage owned by the macro file.

If we have copy file included in a macro file, then in the outline of the macro file we want to have all the variable symbols of the copy file under one node.
So for each var symbol in the vardef storage of the macro file I have to first decide whether it is owned by the macro file or by some copy file (that is the problem). If it is owned by copy file I have to then create node and add the variable symbol under the node.

The same for sequence symbols.

The solution for variable and sequence symbols is not really good. If there is some better solution, then I did not find it. If the copy files and their variable / sequence symbols are planned to be redone at some point, then it might be valid to delete them from the outline for now and have the outline only for the ordinary symbols.

Though that decision is for you and @michalbali256 

Then again, it is also possible that I just implemented the worst possible solution.